document-alerting-on-system-data (#2599)
giantswarm/docs: src/content/tutorials/observability/alerting/create-rules/_index.md

  - https://github.com/orgs/giantswarm/teams/team-atlas
---

> The Giant Swarm Observability Platform provides an [alerting pipeline]({{< relref "/overview/observability/alerting/" >}}) that you can [configure per tenant]({{< relref "/tutorials/observability/alerting/configure-alertmanager" >}}). This platform allows you to create your own alerting and recording rules per tenant.

Reviewer: ```suggestion
The Giant Swarm Observability Platform provides an [alerting pipeline]({{< relref "/overview/observability/alerting/" >}}) that you can [configure per tenant]({{< relref "/tutorials/observability/alerting/configure-alertmanager" >}}). This enables you to create your own alerting and recording rules per tenant.
```

Following Giant Swarm's GitOps approach, you define alerting and recording rules using [Prometheus Operator](https://prometheus-operator.dev/) `PrometheusRule` resources. You can deploy these rules to both management clusters and workload clusters.

> The `observability.giantswarm.io/tenant` label on your rules must reference an existing tenant defined in a [Grafana Organization]({{< relref "/tutorials/observability/multi-tenancy/creating-grafana-organization/" >}}). The system ignores any `PrometheusRule` that references a non-existing tenant. Learn more about our multi-tenancy in [Multi-tenancy in the observability platform]({{< relref "/tutorials/observability/multi-tenancy/" >}}).

Reviewer: ```suggestion
**Warning:** The `observability.giantswarm.io/tenant` label on your rules must reference an existing tenant defined in a [Grafana Organization]({{< relref "/tutorials/observability/multi-tenancy/creating-grafana-organization/" >}}). The system ignores any `PrometheusRule` that references a non-existing tenant. Learn more about our multi-tenancy in [Multi-tenancy in the observability platform]({{< relref "/tutorials/observability/multi-tenancy/" >}}).
```

        for: 5m
        labels:
          # Severity level for routing and prioritization
          severity: critical
```

### Key components

- **`alert`**: unique name for the alert rule
- **`expr`**: PromQL or LogQL expression that defines when the alert should fire
- **`for`**: duration the condition must be true before the alert fires
- **`labels`**: key-value pairs for routing and grouping alerts
- **`annotations`**: human-readable information about the alert

For guidance on writing effective PromQL queries, refer to the [Prometheus querying documentation](https://prometheus.io/docs/prometheus/latest/querying/basics/). You can also explore queries in your [installation's Grafana]({{< relref "/tutorials/observability/data-exploration/accessing-grafana" >}}) explore interface.

## Create recording rules

[Recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/) let you pre-compute frequently needed or computationally expensive expressions. They save results as new time series. This improves query performance and enables the creation of custom metrics for dashboards and alerts.

### When to use recording rules

> Recording rules are useful when you need to:

Reviewer: ```suggestion
[Recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/#recording-rules) are useful when you need to:
```

          # Optional: link to troubleshooting documentation
          runbook_url: https://my-runbook-url
        # PromQL expression that defines the alert condition
        expr: up{job=~"component/.*"} == 0
        # Duration the condition must be true before firing
        for: 5m
        labels:
          # Severity level for routing and prioritization
          severity: critical
```

### Key components

- **`alert`**: unique name for the alert rule
- **`expr`**: PromQL or LogQL expression that defines when the alert should fire
- **`for`**: duration the condition must be true before the alert fires
- **`labels`**: key-value pairs for routing and grouping alerts
- **`annotations`**: human-readable information about the alert

{{< admonition type="tip" title="PromQL Resources" >}}
For guidance on writing effective PromQL queries, refer to the [Prometheus querying documentation](https://prometheus.io/docs/prometheus/latest/querying/basics/). You can also explore queries in your [installation's Grafana]({{< relref "/tutorials/observability/data-exploration/accessing-grafana" >}}) explore interface.
{{< /admonition >}}

## Create recording rules

> [Recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/) allow you to precompute frequently needed or computationally expensive expressions and save their results as new time series. This improves query performance and enables the creation of custom metrics for dashboards and alerts.

Author: ```suggestion
[Recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/) allow you to pre-compute frequently needed or computationally expensive expressions and save their results as new time series. This improves query performance and enables the creation of custom metrics for dashboards and alerts.
```

## Rule scoping

The Observability Platform provides scoping mechanisms to prevent conflicts when deploying the same rules across multiple clusters within a tenant.

### Scoping behavior

#### Workload cluster deployment (cluster-scoped)

When you deploy a `PrometheusRule` in a workload cluster, the system automatically scopes the rules to that specific cluster. For example, deploying a rule with expression `up{job="good"} > 0` in workload cluster `alpha1` results in the loaded expression: `up{cluster_id="alpha1", job="good"} > 0`.

#### Management cluster deployment (installation-scoped)

When you deploy a `PrometheusRule` in a management cluster, the rules target all clusters in the installation without modification.

### Limitations

- The system only provides cluster scoping for metric-based alerts due to upstream limitations
- Scoping applies when teams deploy the same rule across multiple clusters
- For rules you deploy per application in different namespaces, you must manage conflicts manually

For multi-environment deployments, consider using unique naming or namespace-specific labeling to avoid conflicts.

## Using tenant federation for system data alerts

> With the introduction of Alloy 1.9, the Giant Swarm Observability Platform supports tenant federation capabilities. These capabilities let you create alerting rules based on system data without duplicating data intake. This feature enables you to reference metrics and logs from the `giantswarm` tenant directly in your own tenant's alerting rules. You use the `monitoring.grafana.com/source_tenants` label for this purpose.

Author: Yes that's why i'm talking about system data but we probably want to talk about tenant fédération as a feature and how to access system (giant swarm data) as an example
